converter/grpc_interceptor.go

```diff
@@ -0,0 +1,268 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package converter
+
+import (
+	"context"
+
+	"go.temporal.io/api/command/v1"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/failure/v1"
+	historypb "go.temporal.io/api/history/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/grpc"
+)
+
+type serviceInterceptor struct {
+	encoders []PayloadEncoder
+}
+
+func (s *serviceInterceptor) encodePayloads(payloads ...*commonpb.Payloads) error {
+	for _, p := range payloads {
+		for _, payload := range p.GetPayloads() {
+			for i := len(s.encoders) - 1; i >= 0; i-- {
+				if err := s.encoders[i].Encode(payload); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) decodePayloads(payloads ...*commonpb.Payloads) error {
+	for _, p := range payloads {
+		for _, payload := range p.GetPayloads() {
+			for _, encoder := range s.encoders {
+				if err := encoder.Decode(payload); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processRequest(req interface{}) error {
+	switch r := req.(type) {
+	case *workflowservice.StartWorkflowExecutionRequest:
+		return s.encodePayloads(r.Input)
+	case *workflowservice.SignalWorkflowExecutionRequest:
+		return s.encodePayloads(r.Input)
+	case *workflowservice.SignalWithStartWorkflowExecutionRequest:
+		return s.encodePayloads(r.Input, r.SignalInput)
+	case *workflowservice.TerminateWorkflowExecutionRequest:
+		return s.encodePayloads(r.Details)
+	case *workflowservice.RespondActivityTaskCompletedRequest:
+		return s.encodePayloads(r.Result)
+	case *workflowservice.RespondActivityTaskCompletedByIdRequest:
+		return s.encodePayloads(r.Result)
+	case *workflowservice.RespondActivityTaskCanceledRequest:
+		return s.encodePayloads(r.Details)
+	case *workflowservice.RespondActivityTaskCanceledByIdRequest:
+		return s.encodePayloads(r.Details)
+	case *workflowservice.RespondActivityTaskFailedRequest:
+		return s.encodeFailure(r.Failure)
+	case *workflowservice.RespondActivityTaskFailedByIdRequest:
+		return s.encodeFailure(r.Failure)
+	case *workflowservice.RespondWorkflowTaskCompletedRequest:
+		return s.processCommands(r.Commands)
+	case *workflowservice.RespondWorkflowTaskFailedRequest:
+		return s.encodeFailure(r.Failure)
+	case *workflowservice.RecordActivityTaskHeartbeatRequest:
+		return s.encodePayloads(r.Details)
+	case *workflowservice.RecordActivityTaskHeartbeatByIdRequest:
+		return s.encodePayloads(r.Details)
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processResponse(response interface{}) error {
+	switch r := response.(type) {
```

Comment:
`QueryWorkflowResponse` is missing here.
Anything with a header is missing too.